src/transport/virtiofs/mod.rs

```diff
@@ -99,19 +103,14 @@ pub type Result<T> = std::result::Result<T, Error>;
 
 impl std::error::Error for Error {}
 
-impl<'a, M: GuestMemory> Reader<'a, M> {
+impl<'a> Reader<'a> {
     /// Construct a new Reader wrapper over `desc_chain`.
-    #[allow(clippy::new_ret_no_self)]
-    pub fn new(mem: &'a M, desc_chain: DescriptorChain<'a, M>) -> Result<Reader<'a, MS<'a, M>>> {
+    pub fn new(
+        mem: &'a GuestMemoryMmap,
```

Comment:
Why hard-coding to "GuestMemoryMmap" and "GuestMemoryAtomic<GuestMemoryMmap>" here?

Reply:
I'm getting 
```
the trait `vm_memory::GuestMemory` is not implemented for `GuestMemoryAtomic<GuestMemoryMmap>`
```
although `GuestMemoryMmap` implements `vm_memory::GuestMemory`

There seems to miss an abstraction of `vm_memory::GuestMemory` for `GuestMemoryAtomic` in vm-memory crate.

Besides, this drops the generic type requirement on Reader/Writer and makes caller's life much easier. Is there need to create a generic type based Reader/Writer? The `VhostUserBackend` trait is based on `GuestMemoryAtomic<GuestMemoryMmap>` as well so I'm using the same concrete type.

Reply:
It should works with vhost-user-backend, not sure whether it breaks rund or not.

Reply:
The prototype of Reader::new()/Writer::new() is not well-formed, so cause the compilation failure.

Reply:
I tried to change it to 
```
 104 impl<'a, M: GuestMemory> Reader<'a, M> {
 105     /// Construct a new Reader wrapper over `desc_chain`.
 106     pub fn new(
 107         mem: &'a M,
 108         desc_chain: DescriptorChain<&M>,
 109     ) -> Result<Reader<'a, MS<'a, M>>> {
```
and got the above errors with adapting nydusd to use it. I think vm-memory should provide a wrapper for GuestMemoryAtomic over GuestMemory trait.

Reply:
We should use vm-memory in another way, will try to solve it later.